### ISSUE TITLE
Feature Request: Web Version

### ISSUE THREAD

**Issue by danieldaeschle** [fabiospampinato/notable#83]
It would be nice to have a PWA Web App which also works offline on mobile phones.

**Comment by vorg** [comment]
Where would the storage be? Currently Notable depends on your HDD (free) and syncs via e.g. Dropbox (free-ish).

**Comment by felixlinker** [comment]
Shouldn't it be possible to link your cloud service provider to the PWA? I know that [www.draw.io](https://github.com/jgraph/drawio) does this so it must be possible in some way.

**Comment by fabiospampinato** [comment]
@vorg you raised a good point, I've kind of talked about this [here](https://github.com/fabiospampinato/notable/issues/135). 

More specifically about the web app though I think that it will have some kind of paid plan for businesses or users who need non-trivial amount of space.

@felixlinker that's an area I will have to explore really, but there's probably a way to do something like that in our web app too.

**Comment by j-f1** [comment]
You could use GitHub for sync — create a private repo under their account, then automatically clone/pull it to clients using [`isomorphic-git`](https://github.com/isomorphic-git/isomorphic-git). This also has the advantage of providing a handy web UI for creating/editing notes.

**Comment by felixlinker** [comment]
@j-f1 I'd say this would be overkill for many "real-world" users which don't have a clue what a version control system is.

**Comment by slikts** [comment]
Google Drive, Dropbox, GitHub Gists and other services offer APIs to synchronize files, so a PWA version would work fine. See StackEdit for a working example. This app doesn't seem to do anything that would warrant it using Electron. A PWA would essentially be an another tab in the browser, while an Electron app runs a whole separate Chromium instance, so it takes up more resources.

**Comment by fabiospampinato** [comment]
> This app doesn't seem to do anything that would warrant it using Electron.

Except it currently needs access to the filesystem, it provides native menus, its shortcuts don't conflict with the browser's, eventually it will provide global shortcuts, it's fullscreen-able etc.

**Comment by slikts** [comment]
Browsers can use fullscreen as well. Synchronization with cloud storage would work fine without the filesystem. Most browser shortcuts can be overridden or alternatives used. Native menus or an "about" window are odd features to single out.

**Comment by fabiospampinato** [comment]
> This app _doesn't seem to do anything_ that would warrant it using Electron

> Synchronization with cloud storage _would_ work fine without the filesystem.

First you mentioned the present time, later you mentioned an hypothetical scenario, I just highlighted why the first sentence was wrong, I'm not saying some things couldn't be implemented in some other way.

**Comment by slikts** [comment]
Electron has a memory overhead, is harder to distribute, and gives very broad permissions to the app, and these tradeoffs should be justified.

**Comment by fabiospampinato** [comment]
@slikts nobody is forcing you do use any app you don't like. Even just for an icon in my dock and a proper menu I prefer using having Notable packaged inside Electron.

**Comment by slikts** [comment]
Chrome 72+ allows installing PWAs in macOS, i.e., they show up as a separate app in Dock.

Menus are a fairly dated UI idiom; a command palette is generally more convenient in normal use.

**Comment by danieldaeschle** [comment]
I would also prefer to have a web app (and electron app optional) because in my company i don't want to install programs.
Also i can't sync my files because i mustn't install apps like google drive or nextcloud sync.

**Comment by jkaberg** [comment]
If one likes to run stuff in a window this is a nice trick, `/path/to/chrome --app=https://path.to.app`

Anyhow as a @danieldaeschle pointed out, running Notable in a browser (and thus as a server) really mitigates the need for sync, any executable etc. In my eyes quite the advantage as one doesn't need to install (and update!) the app everywhere one uses Notable.

**Comment by danieldaeschle** [comment]
If the main developer don't want to implement this we should consider if  someone start a fork.

**Comment by fabiospampinato** [comment]
@slikts obviously we can't support Chrome 72+ only, plus those apps feel less native than Electron ones I think, i.e. you can't embed the semaphore (the close, minimize and maximize buttons) into the app.

> running Notable in a browser (and thus as a server) really mitigates the need for sync

@jkaberg I'm not sure what you mean. One of Notable's strengths is that notes are plain markdown files, if everything is only online that advantage is gone. Plus you want to make sure everything is accessible even without an internet connection (we can't just cache all your attachments into the PWA).

I'm actually interested in making this happen, I think this is a main component of Notable's long term sustainability, as stated [here](https://github.com/fabiospampinato/notable/issues/135) and [here](https://github.com/fabiospampinato/notable/issues/353).

**Comment by slikts** [comment]
PWAs work prior to Chrome 72 and in other browsers as well; Chrome 72 is what enables installation on macOS specifically, but they still work without installing, and they're installable in earlier Chrome versions on other platforms.

It would still work on plain markdown files; see how StackEdit works. Browser storage quotas are large also enough to hold gigabytes of media files.

**Comment by jkaberg** [comment]
@fabiospampinato being markdown files saved at the web server is perfectly fine, and an advantage if you ask me (proprietary formats and all that).  What I mean by mitigating the sync option is that if Notable is truly responsive running in any browser one doesn't really need sync as long as you can access Notable via web (and thus don't need specific device clients that sync).

I love that you put this on the road map, and will contribute money towards it.

**Comment by j-f1** [comment]
@slikts Are you sure there’s enough space to store gigabytes on the client side? I haven’t been able to find an API that would allow a site to permanently cache that much.

**Comment by fabiospampinato** [comment]
> PWAs work prior to Chrome 72 and in other browsers as well;

Yes, but no Dock icon then. I'm just saying the Electron app isn't going anywhere.

> Browser storage quotas are also large enough to hold gigabytes of media files.

I doubt that. First of all people might not have gigabytes of storage available in their smartphone for instance, and even if they had we can't just waste gigabytes of bandwidth on this.

**Comment by slikts** [comment]
@j-f1 [Firefox storage limits](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Browser_storage_limits_and_eviction_criteria#Storage_limits) and [Chrome storage limits](https://developer.chrome.com/apps/offline_storage#sharing-the%20pool). I've stored gigabytes of media in IndexedDB before.

> Yes, but no Dock icon then.

Chrome 72 will be released in the stable channel soon enough.

> I doubt that. First of all people might not have gigabytes of storage available in their smartphone for instance …

It's not like Electron would run on mobile devices either; moreover, since cloud storage is used, only the files that get accessed need to be synced.

**Comment by MelBourbon** [comment]
I mean, currently I sync my notes with my nextcloud instance to use notable on my Windows and Linux PC's. I'm also able to edit them with the [nextcloud/notes](https://github.com/nextcloud/notes) app but without the nice interface and missing functionality (tags, notebooks etc.). So this would be one option to use a similar system architecture even if nextcloud is much more than needed in this case,

**Comment by danieldaeschle** [comment]
My favorite solutions would be:
1. selfhosted
2. paid hosted

**Comment by nsheff** [comment]
To continue this discussion, would it make sense to integrate in some way with [hedgedoc](hedgedoc.org), which provides a web app and remote server for collaborative, real-time editing of markdown documents?

I love both applications and see the utility of both. Combining their strengths leads to some interesting ideas.

I like the fact that notable can be used offline, and notes are native markdown files. I like the knowing I'm not locked in to a server or database format. I like the tag system and menus to organize these local notes. And personally, I like the dedicated electron app that feels separate from yet another tab in my already overloaded chrome browser windows.

But for notes I want to share,I also like the server approach of hedgedoc, which provides remote syncing, and, more importantly, real-time collaborative editing.  I like that I can edit with collaborators in real time, and share using a link. I like the idea of group-level permission control, and that API calls can pipe remote markdown documents into other tools, like pandoc.

A combination of the two would be very powerful. It's a notable interface that stores notes locally, but could also view/edit remote notes from a hedgedoc server; these remote notes would live alongside my local notes and be taggable with my notable tags. And for real-time collaboration, the notable editor could behave like hedgedoc, showing multiple users editing notes simultaneously. I could "share" one of my local notes, moving it into the hedgedoc database to mint a link and make it collaborative.
